docs/man/getmac.1
.\" Man page generated from reStructuredText.
.
.TH "GETMAC" "1" "Oct 06, 2018" "0.6.0" "getmac"
.SH NAME
getmac \- Cross-platform Python package to get MAC addresses
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
\fI\%Latest version on PyPI\fP
\fI\%Documentation status\fP
\fI\%Travis CI build status\fP
\fI\%Appveyor build status\fP
.sp
Get the MAC address of remote LAN hosts and system network interfaces using Python.
.sp
It provides a platform\-independant interface to get the MAC addresses of:
.INDENT 0.0
.IP \(bu 2
System network interfaces (by interface name)
.IP \(bu 2
Remote hosts on the local network (by IPv4/IPv6 address or hostname)
.UNINDENT
.sp
It provides one function: \fBget_mac_address()\fP
.sp
\fI\%asciicast\fP
.sp
\fI\%asciicast\fP
.INDENT 0.0
.IP \(bu 2
Pure\-Python
.IP \(bu 2
Supports Python 2.6+, 3.4+, pypy, and pypy3
.IP \(bu 2
No dependencies
.IP \(bu 2
Small size
.IP \(bu 2
Can be used as an independent .py file
.IP \(bu 2
Simple terminal tool (when installed as a package)
.UNINDENT
.SH PYTHON EXAMPLES
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
from getmac import get_mac_address
eth_mac = get_mac_address(interface="eth0")
win_mac = get_mac_address(interface="Ethernet 3")
ip_mac = get_mac_address(ip="192.168.0.1")
ip6_mac = get_mac_address(ip6="::1")
host_mac = get_mac_address(hostname="localhost")
updated_mac = get_mac_address(ip="10.0.0.1", network_request=True)

# Enabling debugging
from getmac import getmac
getmac.DEBUG = 2  # DEBUG level 2
print(getmac.get_mac_address(interface="Ethernet 3"))

# Changing the port used for updating ARP table (UDP packet)
from getmac import getmac
getmac.PORT = 44444  # Default: 55555
print(get_mac_address(ip="192.168.0.1", network_request=True))

.ft P
.fi
.UNINDENT
.UNINDENT
.SH TERMINAL EXAMPLES
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
getmac \-\-help
getmac \-\-version

# No arguments will return MAC of the default interface.
getmac
python \-m getmac

# Interface names, IPv4/IPv6 addresses, or Hostnames can be specified
getmac \-\-interface ens33
getmac \-\-ip 192.168.0.1
getmac \-\-ip6 ::1
getmac \-\-hostname home.router

# Running as a Python module with shorthands for the arguments
python \-m getmac \-i \(aqEthernet 4\(aq
python \-m getmac \-4 192.168.0.1
python \-m getmac \-6 ::1
python \-m getmac \-n home.router

# Getting the MAC address of a remote host obviously requires
# the ARP table to be populated. By default, getmac will do
# this for you by sending a small UDP packet to a high port (55555)
# If you don\(aqt want this to happen, you can disable it.
# This is useful if you\(aqre 100% certain the ARP table will be
# populated already, or in red team/forensic scenarios.
getmac \-\-no\-network\-request \-4 192.168.0.1
python \-m getmac \-\-no\-network\-request \-n home.router

# Debug levels can be specified with \(aq\-d\(aq
getmac \-\-debug
python \-m getmac \-d \-i enp11s4
python \-m getmac \-dd \-n home.router

.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Note: the terminal interface will not work on Python 2.6 and older (Sorry RHEL 6 users!).
.SH GET_MAC_ADDRESS()
.INDENT 0.0
.IP \(bu 2
\fBinterface\fP: Name of a network interface on the system.
.IP \(bu 2
\fBip\fP: IPv4 address of a remote host.
.IP \(bu 2
\fBip6\fP: IPv6 address of a remote host.
.IP \(bu 2
\fBhostname\fP: Hostname of a remote host.
.IP \(bu 2
\fBnetwork_request\fP: If an network request should be made to update
and populate the ARP/NDP table of remote hosts used to lookup MACs
in most circumstances. Disable this if you want to just use what\(aqs
already in the table, or if you have requirements to prevent network
traffic. The network request is a empty UDP packet sent to a high
port, 55555 by default. This can be changed by setting \fBgetmac.PORT\fP
to the desired integer value. Additionally, on Windows, this will
send a UDP packet to 1.1.1.1:53 to attempt to determine the default interface.
.UNINDENT
.SH NOTES
.INDENT 0.0
.IP \(bu 2
If none of the arguments are selected, the default
network interface for the system will be used.
.IP \(bu 2
"Remote hosts" refer to hosts in your local layer 2 network, also
commonly referred to as a "broadcast domain", "LAN", or "VLAN". As far
as I know, there is not a reliable method to get a MAC address for a
remote host external to the LAN. If you know any methods otherwise, please
open a GitHub issue or shoot me an email, I\(aqd love to be wrong about this.
.IP \(bu 2
The first four arguments are mutually exclusive. \fBnetwork_request\fP
does not have any functionality when the \fBinterface\fP argument is
specified, and can be safely set if using in a script.
.IP \(bu 2
The physical transport is assumed to be Ethernet (802.3). Others, such as
Wi\-Fi (802.11), are currently not tested or considored. I plan to
address this in the future, and am definitely open to pull requests
or issues related to this, including error reports.
.IP \(bu 2
Exceptions will be handled silently and returned as a None.
If you run into problems, you can set DEBUG to true and get more
information about what\(aqs happening. If you\(aqre still having issues,
please create an issue on GitHub and include the output with DEBUG enabled.
.IP \(bu 2
Messages are output using the \fBwarnings\fP module, and \fBprint()\fP if
\fBgetmac.DEBUG\fP enabled (any value greater than 0).
If you are using logging, they can be captured using logging.captureWarnings().
Otherwise, they can be suppressed using warnings.filterwarnings("ignore").
https://docs.python.org/3/library/warnings.html
.UNINDENT
.SH DOCKER EXAMPLES
.sp
Run getmac container and provide flags
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
docker run \-it ghostofgoes/getmac:latest \-\-help
docker run \-it ghostofgoes/getmac:latest \-\-version
docker run \-it ghostofgoes/getmac:latest \-n localhost

.ft P
.fi
.UNINDENT
.UNINDENT
.sp
There is a pre\-built container located on the
\fI\%Docker hub\fP\&.
Alternatively, you can build the image yourself (from the repository root directory):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
docker build . \-t getmac

.ft P
.fi
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
Windows
.INDENT 2.0
.IP \(bu 2
Commands: \fBgetmac\fP, \fBipconfig\fP
.IP \(bu 2
Libraries: \fBuuid\fP, \fBctypes\fP
.IP \(bu 2
Third\-party Packages: \fBnetifaces\fP, \fBpsutil\fP, \fBscapy\fP
.UNINDENT
.IP \(bu 2
Linux/Unix
.INDENT 2.0
.IP \(bu 2
Commands: \fBarp\fP, \fBip\fP, \fBifconfig\fP, \fBnetstat\fP, \fBip link\fP
.IP \(bu 2
Libraries: \fBuuid\fP, \fBfcntl\fP
.IP \(bu 2
Third\-party Packages:  \fBnetifaces\fP, \fBpsutil\fP, \fBscapy\fP, \fBarping\fP
.IP \(bu 2
Default interfaces: \fBroute\fP, \fBip route list\fP
.IP \(bu 2
Files: \fB/sys/class/net/X/address\fP, \fB/proc/net/arp\fP
.UNINDENT
.IP \(bu 2
Mac OSX (Darwin)
.INDENT 2.0
.IP \(bu 2
\fBnetworksetup\fP
.IP \(bu 2
Same commands as Linux
.UNINDENT
.IP \(bu 2
WSL: Windows commands are used for remote hosts,
and Unix commands are used for interfaces
.UNINDENT
.sp
All or almost all features should work on "supported" platforms (OSes).
.INDENT 0.0
.IP \(bu 2
Windows
.INDENT 2.0
.IP \(bu 2
Desktop: 7, 8, 8.1, 10
.IP \(bu 2
Server: TBD
.IP \(bu 2
(Partially supported, untested): 2000, XP, Vista
.UNINDENT
.IP \(bu 2
Linux distros
.INDENT 2.0
.IP \(bu 2
CentOS/RHEL 6+
.IP \(bu 2
Ubuntu 16+ (14 and older should work as well)
.IP \(bu 2
Fedora
.UNINDENT
.IP \(bu 2
MacOSX (Darwin)
.INDENT 2.0
.IP \(bu 2
The latest two versions probably (TBD)
.UNINDENT
.IP \(bu 2
Windows Subsystem for Linux (WSL)
.IP \(bu 2
Docker
.UNINDENT
.sp
All sub\-versions are the latest available on your platform (with the exception of 2.6).
.INDENT 0.0
.IP \(bu 2
2.6.6 (CentOS 6/RHEL 6 version)
.IP \(bu 2
2.7
.IP \(bu 2
3.4
.IP \(bu 2
3.5
.IP \(bu 2
3.6
.IP \(bu 2
3.7
.UNINDENT
.sp
Please report any problems by opening a issue on GitHub!
.SH CAVEATS
.INDENT 0.0
.IP \(bu 2
Depending on the platform, there could be a performance detriment,
due to heavy usage of regular expressions.
.IP \(bu 2
Platform test coverage is imperfect. If you\(aqre having issues,
then you might be using a platform I haven\(aqt been able to test.
Keep calm, open a GitHub issue, and I\(aqd be more than happy to help.
.IP \(bu 2
Older Python versions (2.5/3.3 and older) are not officially supported.
If you\(aqre running these, all is not lost! Simply copy/paste \fBgetmac.py\fP
into your codebase and make the necessary edits to be compatible with
your version and distribution of Python.
.UNINDENT
.SH KNOWN ISSUES
.INDENT 0.0
.IP \(bu 2
Hostnames for IPv6 devices are not yet supported.
.IP \(bu 2
Windows: the "default" of selecting the default route interface only
works effectively if \fBnetwork_request\fP is enabled. Otherwise,
\fBEthernet\fP as the default.
.IP \(bu 2
There is are currently no automated tests for Python 2.6, which means
there is a much higher potential for regressions. Open an issue if you
encounter any.
.UNINDENT
.sp
Contributors are more than welcome!
See the \fI\%contribution guide\fP to get started,
and checkout the \fI\%todo list\fP for a full list of tasks and bugs.
.sp
Before submitting a PR, please make sure you\(aqve completed the
\fI\%pull request checklist\fP!
.sp
The \fI\%Python Discord server\fP is a good place
to ask questions or discuss the project (Handle: @KnownError).
.SH CONTRIBUTORS
.INDENT 0.0
.IP \(bu 2
Christopher Goes (@ghostofgoes) \- Author and maintainer
.IP \(bu 2
Calvin Tran (@cyberhobbes) \- Windows interface detection improvements
.IP \(bu 2
Izra Faturrahman (@Frizz925) \- Unit tests using the platform samples
.IP \(bu 2
Jose Gonzalez (@Komish) \- Docker container and Docker testing
.IP \(bu 2
@fortunate\-man \- Awesome usage videos
.IP \(bu 2
@martmists \- legacy Python compatibility improvements
.UNINDENT
.sp
Many of the methods used to acquire an address and the core logic framework
are attributed to the CPython project\(aqs UUID implementation.
.INDENT 0.0
.IP \(bu 2
https://github.com/python/cpython/blob/master/Lib/uuid.py
.IP \(bu 2
https://github.com/python/cpython/blob/2.7/Lib/uuid.py
.UNINDENT
.SH OTHER NOTABLE SOURCES
.INDENT 0.0
.IP \(bu 2
\fI\%_unix_fcntl_by_interface\fP
.IP \(bu 2
\fI\%_windows_get_remote_mac_ctypes\fP
.IP \(bu 2
\fI\%String joining\fP
.UNINDENT
.sp
MIT. Feel free to copy, modify, and use to your heart\(aqs content. Enjoy :)
.SH AUTHOR
Christopher Goes
.SH COPYRIGHT
2018, Christopher Goes
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# DEV (xx/xx/2018)

## Dev
* Added unit tests for the samples (Credit: @Frizz925)

## Added
* scripts for building RPMs in the /scripts directory (Credit: @hargoniX)
* pip2 will install an executor named getmac2 and pip3 an executor named getmac so they don't conflict when both RPMs are installed on the same system (Credit: @hargoniX)

# 0.6.0 (10/06/2018)

## Added
* Windows default interface detection if `network_request` is enabled (Credit: @cyberhobbes)
* Docker container (Credit: @Komish)

## Changed
* Changed name to `getmac`. This applies to everything, including
command line tool, PyPI, GitHub, and the documentation.
This is a breaking change, but needed to happen to remove
a huge amount of ambiguity that was causing issues with packaging,
documentation, and several other efforts, not to mention my sanity.
Long-term, the only downside is a conflict on Windows CLI with `getmac.exe`.
* Use proper Python 2-compatible print functions (Credit: @martmists)

## Removed
* Support for Python 2.5. It is not feasible to test, and potentially
breaks some useful language features, such as `__future__`
* Variables PORT and DEBUG from top-level package imports, since changing
them would have no actual effect on execution. Instead, use `getmac.getmac.DEBUG`.

## Dev
* Added example videos demonstrating usage (Credit: @fortunate-man)
* Added contribution guide
* Added documentation on ReadTheDocs
* Added a manpage


# 0.5.0 (09/24/2018)
## Added
* Full support for Windows Subsystem for Linux (WSL). This is working for
all features, including default interface selection! The only edge case
is lookup of remote host IP addresses that are actually local interfaces
won't resolve to a MAC (which should be ff-ff-ff-ff-ff-ff).
## Changed
* Require `argparse` if Python version is 2.6 or older
## Dev
* Updated tox tests: added Jython and IronPython, removed 2.6


# 0.4.0 (09/21/2018)
## Added
* New methods for remote host MACs
    * Windows: `arp`
    * POSIX: `arpreq` package
* New methods for interface MACs
    * Windows: `wmic nic`
* DEBUG levels: DEBUG value is now an integer, and increasing it will
increase the amount and verbosity of output. On the CLI, it can be
configured by increasing the amount of characters for the debug argument,
e.g. '-dd' for DEBUG level 2.
* Jython support (Note: on Windows Jython currently only works with interfaces)
* IronPython support

## Changed
* **Significant** performance improvement for remote hosts. Previously,
the average for `get_mac_address(ip='10.0.0.100')` was 1.71 seconds.
Now, the average is `12.7 miliseconds`, with the special case of a unpopulated
arp table being only slightly higher. This was brought about by changes in
how the arp table is populated. The original method was to use the
host's `ping` command to send an ICMP packet to the host. This took time,
which heavily delayed the ability to actually get an address. The solution
is to instead simply send a empty UDP packet to a high port. The port
this packet is sent to can be configured using the module variable `getmac.PORT`.
* "Fixed" resolution of localhost/127.0.0.1 by hardcoding the response.
This should resolve a lot of problematic edge cases. I'm ok with this
for now since I don't know of a case when it isn't all zeroes.
* Greatly increased the reliability of getting host and interface MACs on Windows
* Improved debugging output
* Tightened up the size of `getmac.py`
* Various minor stability and performance improvements
* Add LICENSE to PyPI package

## Removed
* Support for Python 3.2 and 3.3. The total downloads from PyPI with
those versions in August was ~53k and ~407K, respectfully. The majority
of those are likely from automated testing (e.g. TravisCI) and not
actual users. Therefore, I've decided to drop support to simplify
development, especially since before 3.4 the 3.x series was still
very much a  "work in progress".

## Dev
* Added automated tests for Windows using Appveyor
* Tox runner for tests
* Added github.io page
* Improved TravisCI testing


# 0.3.0 (08/30/2018)
## Added
* Attempt to use Python modules if they're installed. This is useful
for larger projects that already have them installed as dependencies,
as they provide a more reliable means of getting information.
    * `psutil`: Interface MACs on all platforms
    * `scapy`: Interface MACs and Remote MACs on all platforms
    * `netifaces`: Interface MACs on Non-Windows platforms
* New methods for remote MACs
    * POSIX: `ip neighbor show`, Abuse of `uuid._arp_getnode()`
* New methods for Interface MACs
    * POSIX: `lanscan -ai` (HP-UX)

## Changed
* Certain critical failures that should never happen will now warn
instead of failing silently.
* Added a sanity check to the `ip6` argument (IPv6 addresses)
* Improved performance in some areas
* Improved debugging output

## Fixed
* Major Bugfix: search of `proc/net/arp` would return shorter addresses in the
same subnet if they came earlier in the sequence. Example: a search for
`192.168.16.2` on Linux would instead return the MAC address of
`192.168.16.254` with no errors or warning whatsoever.
* Significantly improved default interface detection. Default
interfaces are now properly detected on Linux and most other
POSIX platforms with `ip` or `route` commands available, or the
`netifaces` Python module.

## Dev
* Makefile
* Vagrantfile to spin up testing VMs for various platforms using [Vagrant](https://www.vagrantup.com/docs/)
* Added more samples of command output on platforms (Ubuntu 18.04 LTS)


# 0.2.4 (08/26/2018)
## Fixed
* Fixed identification of remote host on OSX
* Resolved hangs and noticeable lag that occurred when "network_request"
was True (the default)


# 0.2.3 (08/07/2018)
## Fixed
* Remote host for Python 3 on Windows


# 0.2.2
## Added
* Short versions of CLI arguments (e.g. "-i" for "--interface")

## Changed
* Improved usage of "ping" across platforms and IP versions
* Various minor tweaks for performance
* Improved Windows detection

## Fixed

getmac/getmac.py
# -*- coding: utf-8 -*-

"""Get the MAC address of remote hosts or network interfaces using Python.

It provides a platform-independant interface to get the MAC addresses of:

* System network interfaces (by interface name)
* Remote hosts on the local network (by IPv4/IPv6 address or hostname)

It provides one function: `get_mac_address()`

Examples:

    from getmac import get_mac_address
    eth_mac = get_mac_address(interface="eth0")
    win_mac = get_mac_address(interface="Ethernet 3")
    ip_mac = get_mac_address(ip="192.168.0.1")
    ip6_mac = get_mac_address(ip6="::1")
    host_mac = get_mac_address(hostname="localhost")
    updated_mac = get_mac_address(ip="10.0.0.1", network_request=True)"""

from __future__ import print_function
import ctypes, os, re, sys, struct, socket, shlex, traceback, platform
from warnings import warn
from subprocess import Popen, PIPE, CalledProcessError
try:
    from subprocess import DEVNULL  # Py3
except ImportError:
    DEVNULL = open(os.devnull, 'wb')  # Py2

__version__ = '0.6.0'
DEBUG = 0
PORT = 55555

PY2 = sys.version_info[0] == 2
_SYST = platform.system()
if _SYST == 'Windows':
    IS_WINDOWS = True
elif _SYST == 'Java':
    IS_WINDOWS = os.sep == '\\'
else:
    IS_WINDOWS = False

if not IS_WINDOWS and 'Microsoft' in platform.version():
    IS_WSL = True
else:
    IS_WSL = False


PATH = os.environ.get('PATH', os.defpath).split(os.pathsep)
if not IS_WINDOWS:
    PATH.extend(('/sbin', '/usr/sbin'))

ENV = dict(os.environ)
ENV['LC_ALL'] = 'C'  # Ensure ASCII output so we parse correctly

IP4 = 0
IP6 = 1
INTERFACE = 2
HOSTNAME = 3

MAC_RE_COLON = r'([0-9a-fA-F]{2}(?::[0-9a-fA-F]{2}){5})'
MAC_RE_DASH = r'([0-9a-fA-F]{2}(?:-[0-9a-fA-F]{2}){5})'
MAC_RE_DARWIN = r'([0-9a-fA-F]{1,2}(?::[0-9a-fA-F]{1,2}){5})'


def get_mac_address(interface=None, ip=None, ip6=None,
                    hostname=None, network_request=True):
    """Get a Unicast IEEE 802 MAC-48 address from a local interface or remote host.

    You must only use one of the first four arguments. If none of the arguments
    are selected, the default network interface for the system will be used.

    Exceptions will be handled silently and returned as a None.
    For the time being, it assumes you are using Ethernet.

    NOTES:
    * You MUST provide str-typed arguments, REGARDLESS of Python version.
    * localhost/127.0.0.1 will always return '00:00:00:00:00:00'

    Args:
        interface (str): Name of a local network interface (e.g "Ethernet 3", "eth0", "ens32")
        ip (str): Canonical dotted decimal IPv4 address of a remote host (e.g 192.168.0.1)
        ip6 (str): Canonical shortened IPv6 address of a remote host (e.g ff02::1:ffe7:7f19)
        hostname (str): DNS hostname of a remote host (e.g "router1.mycorp.com", "localhost")
        network_request (bool): Send a UDP packet to a remote host to populate
        the ARP/NDP tables for IPv4/IPv6. The port this packet is sent to can
        be configured using the module variable `getmac.PORT`.
    Returns:
        Lowercase colon-separated MAC address, or None if one could not be
        found or there was an error."""
    if (hostname and hostname == 'localhost') or (ip and ip == '127.0.0.1'):
        return '00:00:00:00:00:00'

    if DEBUG >= 2:
        print("IS_WINDOWS\t%s\nIS_WSL\t\t%s" % (str(IS_WINDOWS), str(IS_WSL)))

    # Resolve hostname to an IP address
    if hostname:
        ip = socket.gethostbyname(hostname)

    # Populate the ARP table by sending a empty UDP packet to a high port
    if network_request and (ip or ip6):
        try:
            if ip:
                s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                s.sendto(b'', (ip, PORT))
            else:
                s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
                s.sendto(b'', (ip6, PORT))
        except Exception:
            if DEBUG:
                print("ERROR: Failed to send ARP table population packet")
            if DEBUG >= 2:
                traceback.print_exc()

    # Setup the address hunt based on the arguments specified
    if ip6:
        if not socket.has_ipv6:
            warn("Cannot get the MAC address of a IPv6 host: "
                 "IPv6 is not supported on this system", RuntimeWarning)
            return None
        elif ':' not in ip6:
            warn("Invalid IPv6 address: %s" % ip6, RuntimeWarning)
            return None
        to_find = ip6
        typ = IP6
    elif ip:
        to_find = ip
        typ = IP4
    else:
        typ = INTERFACE
        if interface:
            to_find = interface
        else:
            # Default to finding MAC of the interface with the default route
            if IS_WINDOWS:
                if network_request:
                    to_find = _fetch_ip_using_dns()
                    typ = IP4
                else:
                    to_find = 'Ethernet'
            else:
                to_find = _hunt_linux_default_iface()
                if not to_find:
                    to_find = 'en0'

    mac = _hunt_for_mac(to_find, typ, net_ok=network_request)
    if DEBUG:
        print("Raw MAC found: %s" % mac)

    # Check and format the result to be lowercase, colon-separated
    if mac is not None:
        mac = str(mac)
        if not PY2:  # Strip bytestring conversion artifacts
            mac = mac.replace("b'", '').replace("'", '')\
                     .replace('\\n', '').replace('\\r', '')
        mac = mac.strip().lower().replace(' ', '').replace('-', ':')

        # Fix cases where there are no colons
        if ':' not in mac and len(mac) == 12:
            if DEBUG:
                print("Adding colons to MAC %s" % mac)
            mac = ':'.join(mac[i:i + 2] for i in range(0, len(mac), 2))

        # Pad single-character octets with a leading zero (e.g Darwin's ARP output)
        elif len(mac) < 17:
            if DEBUG:
                print("Length of MAC %s is %d, padding single-character "
                      "octets with zeros" % (mac, len(mac)))
            parts = mac.split(':')
            new_mac = []
            for part in parts:
                if len(part) == 1:
                    new_mac.append('0' + part)
                else:
                    new_mac.append(part)
            mac = ':'.join(new_mac)

        # MAC address should ALWAYS be 17 characters before being returned
        if len(mac) != 17:
            if DEBUG:
                print("ERROR: MAC %s is not 17 characters long!" % mac)
            mac = None
    return mac


def _search(regex, text, group_index=0):
    match = re.search(regex, text)
    if match:
        return match.groups()[group_index]


def _popen(command, args):
    for directory in PATH:
        executable = os.path.join(directory, command)
        if (os.path.exists(executable) and
                os.access(executable, os.F_OK | os.X_OK) and
                not os.path.isdir(executable)):
            break
    else:
        executable = command
    if DEBUG >= 3:
        print("Running: '%s %s'" % (executable, args))
    return _call_proc(executable, args)


def _call_proc(executable, args):
    if IS_WINDOWS:
        cmd = executable + ' ' + args
    else:
        cmd = [executable] + shlex.split(args)

    # Popen instead of check_output() for Python 2.6 compatibility
    process = Popen(cmd, stdout=PIPE, stderr=DEVNULL, env=ENV)
    output, unused_err = process.communicate()
    retcode = process.poll()
    if retcode:
        raise CalledProcessError(retcode, cmd, output=output)

    if not PY2 and isinstance(output, bytes):
        return str(output, 'utf-8')
    else:
        return str(output)


def _windows_ctypes_host(host):
    if not PY2:  # Convert to bytes on Python 3+ (Fixes #7)
        host = host.encode()
    try:
        inetaddr = ctypes.windll.wsock32.inet_addr(host)
        if inetaddr in (0, -1):
            raise Exception
    except Exception:
        hostip = socket.gethostbyname(host)
        inetaddr = ctypes.windll.wsock32.inet_addr(hostip)

    buffer = ctypes.c_buffer(6)
    addlen = ctypes.c_ulong(ctypes.sizeof(buffer))

    send_arp = ctypes.windll.Iphlpapi.SendARP
    if send_arp(inetaddr, 0, ctypes.byref(buffer), ctypes.byref(addlen)) != 0:
        return None

    # Convert binary data into a string.
    macaddr = ''
    for intval in struct.unpack('BBBBBB', buffer):
        if intval > 15:
            replacestr = '0x'
        else:
            replacestr = 'x'
        macaddr = ''.join([macaddr, hex(intval).replace(replacestr, '')])
    return macaddr


def _fcntl_iface(iface):
    import fcntl
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    # 0x8927 = SIOCGIFADDR
    info = fcntl.ioctl(s.fileno(), 0x8927, struct.pack('256s', iface[:15]))
    return ':'.join(['%02x' % ord(char) for char in info[18:24]])


def _psutil_iface(iface):
    import psutil
    nics = psutil.net_if_addrs()
    if iface in nics:
        nic = nics[iface]
        for i in nic:
            if i.family == psutil.AF_LINK:
                return i.address


def _netifaces_iface(iface):
    """This method does not work on Windows."""
    import netifaces
    return netifaces.ifaddresses(iface)[netifaces.AF_LINK][0]['addr']


def _scapy_ip(ip):
    """Requires root permissions on POSIX platforms.
    Windows does not have this limitation."""
    from scapy.layers.l2 import getmacbyip
    return getmacbyip(ip)


def _scapy_iface(iface):
    from scapy.layers.l2 import get_if_hwaddr
    if IS_WINDOWS:
        from scapy.arch.windows import get_windows_if_list
        interfaces = get_windows_if_list()
        for i in interfaces:
            if any(iface in i[x] for x in ['name', 'netid', 'description', 'win_index']):
                return i['mac']
    # Do not put an 'else' here!
    return get_if_hwaddr(iface)


def _uuid_ip(ip):
    from uuid import _arp_getnode
    backup = socket.gethostbyname
    try:
        socket.gethostbyname = lambda x: ip
        mac1 = _arp_getnode()
        if mac1 is not None:
            mac1 = _uuid_convert(mac1)
            mac2 = _arp_getnode()
            mac2 = _uuid_convert(mac2)
            if mac1 == mac2:
                return mac1
    except Exception:
        raise
    finally:
        socket.gethostbyname = backup


def _uuid_lanscan_iface(iface):
    if not PY2:
        iface = bytes(iface)
    mac = __import__('uuid')._find_mac('lanscan', '-ai', [iface], lambda i: 0)
    if mac:
        return _uuid_convert(mac)


def _uuid_convert(mac):
    return ':'.join(('%012X' % mac)[i:i+2] for i in range(0, 12, 2))


def _hunt_for_mac(to_find, type_of_thing, net_ok=True):
    """Format of method lists:
    Tuple:  (regex, regex index, command, command args)
            Command args is a list of strings to attempt to use as arguments
    lambda: Function to call"""
    if not PY2 and isinstance(to_find, bytes):
        to_find = str(to_find, 'utf-8')

    # Windows - Network Interface
    if IS_WINDOWS and type_of_thing == INTERFACE:
        methods = [
            # getmac - Connection Name
            (r'\r\n' + to_find + r'.*' + MAC_RE_DASH + r'.*\r\n',
             0, 'getmac.exe', ['/NH /V']),

            # ipconfig
            (to_find + r'(?:\n?[^\n]*){1,8}Physical Address[ .:]+' + MAC_RE_DASH + r'\r\n',
             0, 'ipconfig.exe', ['/all']),

            # getmac - Network Adapter (the human-readable name)
            (r'\r\n.*' + to_find + r'.*' + MAC_RE_DASH + r'.*\r\n',
             0, 'getmac.exe', ['/NH /V']),

            # wmic - WMI command line utility
            lambda x: _popen('wmic.exe', 'nic where "NetConnectionID = \'%s\'" get '
                                         'MACAddress /value' % x).strip().partition('=')[2],

            _psutil_iface,
            _scapy_iface,
        ]

    # Windows - Remote Host
    elif (IS_WINDOWS or IS_WSL) and type_of_thing in [IP4, IP6, HOSTNAME]:
        methods = [
            # arp -a - Parsing result with a regex
            (MAC_RE_DASH, 0, 'arp.exe', ['-a %s' % to_find]),

            _scapy_ip,
        ]

        # Add methods that make network requests
        if net_ok and type_of_thing != IP6 and not IS_WSL:
            methods.insert(1, _windows_ctypes_host)

    # Non-Windows - Network Interface
    elif type_of_thing == INTERFACE:
        methods = [
            lambda x: _popen('cat', '/sys/class/net/' + x + '/address'),

            _fcntl_iface,

            # Fast ifconfig
            (r'HWaddr ' + MAC_RE_COLON,
             0, 'ifconfig', [to_find]),

            # Fast Mac OS X
            (r'ether ' + MAC_RE_COLON,
             0, 'ifconfig', [to_find]),

            # netstat
            (to_find + r'.*(HWaddr) ' + MAC_RE_COLON,
             1, 'netstat', ['-iae']),

            # ip link (Don't use 'list' due to SELinux [Android 24+])
            (to_find + r'.*\n.*link/ether ' + MAC_RE_COLON,
             0, 'ip', ['link %s' % to_find, 'link']),

            # Quick attempt on Mac OS X
            (MAC_RE_COLON,
             0, 'networksetup', ['-getmacaddress %s' % to_find]),

            # ifconfig
            (to_find + r'.*(HWaddr) ' + MAC_RE_COLON,
             1, 'ifconfig', ['', '-a', '-v']),

            # Mac OS X
            (to_find + r'.*(ether) ' + MAC_RE_COLON,
             1, 'ifconfig', ['']),

            # Tru64 ('-av')
            (to_find + r'.*(Ether) ' + MAC_RE_COLON,
             1, 'ifconfig', ['-av']),

            _netifaces_iface,
            _psutil_iface,
            _scapy_iface,
            _uuid_lanscan_iface,
        ]

    # Non-Windows - Remote Host
    elif type_of_thing in [IP4, IP6, HOSTNAME]:
        esc = re.escape(to_find)
        methods = [
            # Need a space, otherwise a search for 192.168.16.2
            # will match 192.168.16.254 if it comes first!
            (esc + r' .+' + MAC_RE_COLON,
             0, 'cat', ['/proc/net/arp']),

            lambda x: _popen('ip', 'neighbor show %s' % x)
            .partition(x)[2].partition('lladdr')[2].strip().split()[0],

            # -a: BSD-style format
            # -n: shows numerical addresses
            (r'\(' + esc + r'\)\s+at\s+' + MAC_RE_COLON,
             0, 'arp', [to_find, '-an', '-an %s' % to_find]),

            # Darwin (OSX) oddness
            (r'\(' + esc + r'\)\s+at\s+' + MAC_RE_DARWIN,
             0, 'arp', [to_find, '-a', '-a %s' % to_find]),

            _uuid_ip,
            _scapy_ip,
            lambda x: __import__('arpreq').arpreq(x),
        ]
    else:
        warn("ERROR: reached end of _hunt_for_mac() if-else chain!", RuntimeError)
        return None
    return _try_methods(methods, to_find)


def _try_methods(methods, to_find=None):
    """We try every method and see if it returned a MAC address. If it returns
    None or raises an exception, we continue and try the next method."""
    found = None
    for m in methods:
        try:
            if isinstance(m, tuple):
                for arg in m[3]:  # list(str)
                    if DEBUG:
                        print("Trying: '%s %s'" % (m[2], arg))
                    # Arguments: (regex, _popen(command, arg), regex index)
                    found = _search(m[0], _popen(m[2], arg), m[1])
                    if DEBUG:
                        print("Result: %s\n" % found)
            elif callable(m):
                if DEBUG:
                    print("Trying: '%s' (to_find: '%s')" % (m.__name__, str(to_find)))
                if to_find is not None:
                    found = m(to_find)
                else:
                    found = m()
                if DEBUG:
                    print("Result: %s\n" % found)
        except Exception as ex:
            if DEBUG:
                print("Exception: %s" % str(ex))
            if DEBUG >= 2:
                traceback.print_exc()
            continue
        if found:
            break
    return found


def _hunt_linux_default_iface():
    # NOTE: for now, we check the default interface for WSL using the
    # same methods as POSIX, since those parts of the net stack work fine.

    methods = [
        lambda: _popen('route', '-n').partition('0.0.0.0')[2].partition('\n')[0].split()[-1],
        lambda: _popen('ip', 'route list 0/0').partition('dev')[2].partition('proto')[0].strip(),
        lambda: list(__import__('netifaces').gateways()['default'].values())[0][1],
    ]

    return _try_methods(methods=methods)


def _fetch_ip_using_dns():
    """Use a UDP socket with Cloudflare's DNS to determine
    the IP of the default network interface"""
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(('1.1.1.1', 53))
    ip = s.getsockname()[0]
    return ip


__all__ = ['get_mac_address']
